code/game/machinery/announcement_system.dm

```diff
@@ -133,83 +109,236 @@ GLOBAL_LIST_EMPTY(announcement_systems)
-		if("newhead_toggle")
-			newhead_toggle = !newhead_toggle
-			update_appearance()
-		if("arrivalToggle")
-			arrival_toggle = !arrival_toggle
-			update_appearance()
-		if("node_toggle")
-			announce_research_node = !announce_research_node
-	add_fingerprint(usr)
-
-/obj/machinery/announcement_system/attack_robot(mob/living/silicon/user)
-	. = attack_ai(user)
+				config.announcement_lines_map[param["lineKey"]] = new_message
+				usr.log_message("updated [param["lineKey"]] line in the [config.name] to: [new_message]", LOG_GAME)
 
-/obj/machinery/announcement_system/attack_ai(mob/user)
-	if(!user.can_perform_action(src, ALLOW_SILICON_REACH))
+/obj/machinery/announcement_system/can_interact(mob/user)
+	. = ..()
+	if (!.)
 		return
-	if(machine_stat & BROKEN)
+
+	if (machine_stat & EMPED)
 		to_chat(user, span_warning("[src]'s firmware appears to be malfunctioning!"))
+		if (!isAI(user))	// Deus Ex Machina goes without multitool in his default complectation.
+			to_chat(user, span_warning("However, you can reset it with [EXAMINE_HINT("multitool")], while its [EXAMINE_HINT("panel is open")]!"))
+		return FALSE
+
+/// If AAS can't broadcast message, it shouldn't be picked by randomizer.
+/obj/machinery/announcement_system/proc/has_supported_channels(list/channels)
+	if (!LAZYLEN(channels) || (RADIO_CHANNEL_COMMON in channels))
+		// Okay, I am not proud of this, but I don't want CentCom or Syndie AASs to broadcast on Common.
+		return src.type == /obj/machinery/announcement_system
+	for(var/channel in channels)
+		if(radio.channels[channel])
+			return TRUE
+	return FALSE
+
+/// Can AAS receive request for broadcast from you?
+/obj/machinery/announcement_system/proc/can_be_reached_from(atom/source)
+	if(!source || !istype(source))
+		return TRUE
+	var/turf/source_turf = get_turf(source)
+	if (!source_turf)
+		return TRUE
+	// Keep updated with broadcasting.dm (/datum/signal/subspace/vocal/New)
+	return z in SSmapping.get_connected_levels(source_turf)
+
+/// Compiles the announcement message with the provided variables. Announcement line is optional.
+/obj/machinery/announcement_system/proc/compile_config_message(aas_config_entry_type, list/variables_map, announcement_line, fail_if_disabled=FALSE)
+	var/datum/aas_config_entry/config = locate(aas_config_entry_type) in config_entries
+	if (!config || (fail_if_disabled && !config.enabled))
 		return
-	interact(user)
+	return config.compile_announce(variables_map, announcement_line)
 
-/obj/machinery/announcement_system/proc/act_up() //does funny breakage stuff
-	if(!atom_break()) // if badmins flag this unbreakable or its already broken
+/// Sends a message to the appropriate channels.
+/obj/machinery/announcement_system/proc/broadcast(message, list/channels, command_span = FALSE)
+	use_energy(active_power_usage)
+	if(!LAZYLEN(channels))
+		radio.talk_into(src, message, null, command_span ? list(speech_span, SPAN_COMMAND) : null)
+	else
```

Comment:
Use an early return instead to reduce indentation.

Reply:
Done. And merge conflicts should be resolved